tests/unit/pluginhandler/test_pluginhandler.py

```diff
@@ -824,6 +829,100 @@ def test_prime(self):
+        })
+
+        handler.pull()
+        metadata = handler.get_pull_state().scriptlet_metadata
+        self.assertThat(metadata.get_version(), Equals('override-version'))
+
+    def test_set_version_in_build(self):
+        handler = self.load_part('test_part', part_properties={
+            'override-build': 'snapcraftctl set-version override-version'
+        })
+
+        handler.pull()
+        handler.build()
+        metadata = handler.get_build_state().scriptlet_metadata
+        self.assertThat(metadata.get_version(), Equals('override-version'))
+        self.assertFalse(handler.get_pull_state().scriptlet_metadata)
+
+    def test_set_version_in_stage(self):
+        handler = self.load_part('test_part', part_properties={
+            'override-stage': 'snapcraftctl set-version override-version'
+        })
+
+        handler.pull()
+        handler.build()
+        handler.stage()
+        metadata = handler.get_stage_state().scriptlet_metadata
+        self.assertThat(metadata.get_version(), Equals('override-version'))
+        self.assertFalse(handler.get_pull_state().scriptlet_metadata)
+        self.assertFalse(handler.get_build_state().scriptlet_metadata)
+
+    def test_set_version_in_prime(self):
+        handler = self.load_part('test_part', part_properties={
+            'override-prime': 'snapcraftctl set-version override-version'
+        })
+
+        handler.pull()
+        handler.build()
+        handler.stage()
+        handler.prime()
+        metadata = handler.get_prime_state().scriptlet_metadata
+        self.assertThat(metadata.get_version(), Equals('override-version'))
+        self.assertFalse(handler.get_pull_state().scriptlet_metadata)
+        self.assertFalse(handler.get_build_state().scriptlet_metadata)
+        self.assertFalse(handler.get_stage_state().scriptlet_metadata)
+
+
+class ScriptletSetVersionErrorTestCase(unit.TestCase):
+
+    scriptlet_scenarios = [
+        ('override-pull', {'override_pull': 'snapcraftctl set-version 1'}),
+        ('override-build', {'override_build': 'snapcraftctl set-version 2'}),
+        ('override-stage', {'override_stage': 'snapcraftctl set-version 3'}),
+        ('override-prime', {'override_prime': 'snapcraftctl set-version 4'}),
+    ]
+
+    scenarios = multiply_scenarios(scriptlet_scenarios, scriptlet_scenarios)
+
+    def test_set_version_multiple_times(self):
+        part_properties = {}
+        with contextlib.suppress(AttributeError):
+            part_properties['override-pull'] = self.override_pull
+        with contextlib.suppress(AttributeError):
+            part_properties['override-build'] = self.override_build
+        with contextlib.suppress(AttributeError):
+            part_properties['override-stage'] = self.override_stage
+        with contextlib.suppress(AttributeError):
+            part_properties['override-prime'] = self.override_prime
+
+        # A few of these test cases result in only one of these scriptlets
```

Comment:
This is why I don't like scenarios as much any more :smile:

Reply:
I know, I really feel like there should be a good way to do that :thinking: .